File: Solution/283.移动零/283.移动零.cpp
/*
 * @Descroption: LeetCode 283. 移动零
 * @Author: EmoryHuang
 * @Date: 2021-05-17 16:42:26
 * @解题思路:
 * 双指针，i遍历数组，j记录非0数组长度
 * 如果nums[i]不为 0 则交换
 */
 
class Solution {
   public:
    void moveZeroes(vector<int> &nums) {
        if (nums.size() <= 1) return;
        int i = 0, j = 0;  // i遍历数组，j记录非0数组长度
        while (i < nums.size()) {
            if (nums[i] != 0) {
                swap(nums[i], nums[j]);  // 如果不为0则交换
                j++;
            }
            i++;
        }
    }
};

File: Solution/283.移动零/283.移动零.java
/*
 * @Description: LeetCode 283. 移动零
 * @Author: EmoryHuang
 * @Date: 2023-08-04 10:05:42
 * @Method:
 * 双指针，分别指向当前元素和非零元素末尾
 */

import java.util.*;

class Solution {
    public void moveZeroes(int[] nums) {
        int fast = 0, slow = 0;
        while (fast < nums.length) {
            if (nums[fast] != 0) {
                swap(nums, fast, slow);
                slow++;
            }
            fast++;
        }
    }

    void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}